### PR TITLE
Add OptimismTVLSubmodule

### PR DESCRIPTION
**Pull request by maciekzygmunt**
Resolves L2B-2392

This PR: 
- Initialize TVL module for Optimism
- Add `totalSupply` tokens for Optimism

### CODE REVIEW

File: packages/backend/src/config/Config.ts [hunk 1]
```diff
@@ -71,6 +72,16 @@ export interface ArbitrumTvlConfig {
   readonly arbiscanApiKey: string
   readonly minBlockTimestamp: UnixTime
 }
+export interface OptimismTvlConfig {
+  readonly providerUrl: string
+  readonly optiscanApiKey: string
+  readonly minBlockTimestamp: UnixTime
+}
+export interface BaseTvlConfig {
```
Comment (sz-piotr) [inline packages/backend/src/config/Config.ts:80]: The four configs are exactly the same. We should use a single interface

File: packages/backend/src/peripherals/optimism/multicall/OptimismMulticall.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+/**
+ * @see https://optimistic.etherscan.io/address/0xe295ad71242373c37c5fda7b57f26f9ea1088afe
+ */
+export const OPTIMISM_MULTICALL_BATCH_SIZE = 150
+export const OPTIMISM_MULTICALL_BLOCK = 0
+export const OPTIMISM_MULTICALL_ADDRESS = EthereumAddress(
+  '0xE295aD71242373C37C5FdA7B57F26f9eA1088AFe',
+)
+
+export class OptimismMulticallClient {
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/optimism/multicall/OptimismMulticall.ts:21]: We can probably make the existing MulticallClient take the encoder/decoder as configuration

**Review by sz-piotr — CHANGES_REQUESTED** [review]

File: packages/backend/src/config/Config.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (maciekzygmunt) [reply packages/backend/src/config/Config.ts]: [d212d14](https://github.com/l2beat/l2beat/pull/1889/commits/d212d14dbc62e148169066e557d79f74799d186e)

File: packages/backend/src/config/Config.ts [hunk 3]
```diff
@@ -71,6 +72,16 @@ export interface ArbitrumTvlConfig {
   readonly arbiscanApiKey: string
   readonly minBlockTimestamp: UnixTime
 }
+export interface OptimismTvlConfig {
+  readonly providerUrl: string
+  readonly optiscanApiKey: string
```
[the remark this answers is not in the document]
Comment (maciekzygmunt) [reply packages/backend/src/config/Config.ts]: [d212d14](https://github.com/l2beat/l2beat/pull/1889/commits/d212d14dbc62e148169066e557d79f74799d186e)

File: packages/backend/src/config/Config.ts [hunk 4]
```diff
@@ -58,6 +58,7 @@ export interface TvlConfig {
   readonly coingeckoApiKey: string | undefined
   readonly ethereum: EthereumTvlConfig | false
   readonly arbitrum: ArbitrumTvlConfig | false
+  readonly optimism: OptimismTvlConfig | false
```
[the remark this answers is not in the document]
Comment (maciekzygmunt) [reply packages/backend/src/config/Config.ts]: [d212d14](https://github.com/l2beat/l2beat/pull/1889/commits/d212d14dbc62e148169066e557d79f74799d186e)

File: packages/backend/src/config/config.local.ts [hunk 5]
```diff
@@ -71,6 +72,11 @@ export function getLocalConfig(): Config {
         providerUrl: getEnv('ARBITRUM_PROVIDER_URL'),
         minBlockTimestamp: UnixTime.now().add(-7, 'days').toStartOf('hour'),
       },
+      optimism: optimismTvlEnabled && {
+        optiscanApiKey: getEnv('OPTISCAN_API_KEY'),
```
[the remark this answers is not in the document]
Comment (maciekzygmunt) [reply packages/backend/src/config/config.local.ts]: [d212d14](https://github.com/l2beat/l2beat/pull/1889/commits/d212d14dbc62e148169066e557d79f74799d186e)

**Review by sz-piotr — APPROVED** [review]
Do not merge before the env variables are updated on heroku

File: packages/backend/src/config/Config.ts [hunk 6]
```diff
@@ -56,19 +56,17 @@ export interface TvlConfig {
...
+  readonly ethereum: ChainTvlConfig | false
+  readonly arbitrum: ChainTvlConfig | false
+  readonly optimism: ChainTvlConfig | false
 }
 
-export interface EthereumTvlConfig {
+export interface ChainTvlConfig {
   readonly providerUrl: string
+  readonly providerCallsPerMinute: number
+  readonly networkName: string
```
Comment (antooni) [inline packages/backend/src/config/Config.ts:67]: this one is not needed probably + in the case of base ethers do not have it as a config option

File: packages/backend/src/modules/tvl/ChainTvlSubmodule.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import { MulticallConfigEntry } from '../../peripherals/ethereum/multicall/types'
+import { TvlSubmodule } from '../ApplicationModule'
+import { TvlDatabase } from './types'
+
+export function chainTvlSubmodule(
+  name: string,
+  chainId: ChainId,
+  projectId: ProjectId,
+  chainTvlConfig: ChainTvlConfig | false,
+  tokens: Token[],
```
Comment (antooni) [inline packages/backend/src/modules/tvl/ChainTvlSubmodule.ts:23]: this array should be filtered

File: packages/config/src/layer2s/optimism.ts [hunk 8]
```diff
@@ -37,13 +37,29 @@ const TOKENS: Omit<Token, 'chainId'>[] = [
...
+  {
+    id: AssetId.USDC_ON_OPTIMISM,
+    name: 'USD Coin',
+    coingeckoId: CoingeckoId('usd-coin'),
+    address: EthereumAddress('0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85'),
+    iconUrl:
+      'https://assets.coingecko.com/coins/images/6319/large/USD_Coin_icon.png?1547042389',
+    symbol: 'USDC',
+    decimals: 6,
+    sinceTimestamp: new UnixTime(1668453318),
```
Comment (antooni) [inline packages/config/src/layer2s/optimism.ts:58]: how was this value obtained? via script?

File: packages/backend/src/modules/ApplicationModule.ts [hunk 9]
```diff
@@ -8,6 +8,6 @@ export interface ApplicationModule {
 }
 
 export interface TvlSubmodule {
-  updaters?: AssetUpdater[]
+  assetUpdaters?: AssetUpdater[]
```
Comment (antooni) [inline packages/backend/src/modules/ApplicationModule.ts:11]: Maybe it should be called "reportUpdaters" and the interface should also change to "ReportUpdater"

File: packages/backend/src/modules/tvl/ChainTvlSubmodule.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+import { TotalSupplyUpdater } from '../../core/totalSupply/TotalSupplyUpdater'
+import { EthereumClient } from '../../peripherals/ethereum/EthereumClient'
+import { MulticallClient } from '../../peripherals/ethereum/multicall/MulticallClient'
+import { MulticallConfigEntry } from '../../peripherals/ethereum/multicall/types'
+import { TvlSubmodule } from '../ApplicationModule'
+import { TvlDatabase } from './types'
+
+export function chainTvlSubmodule(
+  name: string,
+  chainId: ChainId,
```
Comment (antooni) [inline packages/backend/src/modules/tvl/ChainTvlSubmodule.ts:20]: this value can be added to ChainTvlConfig

File: packages/backend/src/modules/tvl/ChainTvlSubmodule.ts [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  projectId: ProjectId,
```
Comment (antooni) [inline packages/backend/src/modules/tvl/ChainTvlSubmodule.ts:21]: maybe this value also should be added to ChainTvlConfig, but I am not sure

File: packages/backend/src/modules/tvl/ChainTvlSubmodule.ts (hunk 7 above)
Comment (antooni) [reply packages/backend/src/modules/tvl/ChainTvlSubmodule.ts:23]: maybe this should be added to ChainTvlConfig

File: packages/backend/src/modules/tvl/ChainTvlSubmodule.ts [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+import { TotalSupplyProvider } from '../../core/totalSupply/TotalSupplyProvider'
+import { TotalSupplyUpdater } from '../../core/totalSupply/TotalSupplyUpdater'
+import { EthereumClient } from '../../peripherals/ethereum/EthereumClient'
+import { MulticallClient } from '../../peripherals/ethereum/multicall/MulticallClient'
+import { MulticallConfigEntry } from '../../peripherals/ethereum/multicall/types'
+import { TvlSubmodule } from '../ApplicationModule'
+import { TvlDatabase } from './types'
+
+export function chainTvlSubmodule(
+  name: string,
```
Comment (antooni) [inline packages/backend/src/modules/tvl/ChainTvlSubmodule.ts:19]:
```suggestion
  submoduleName: string,
```

Comment (antooni) [reply packages/backend/src/modules/tvl/ChainTvlSubmodule.ts:19]: or we can skip it alltogether and just use ChainId.getName() to create this string

**Review by antooni — CHANGES_REQUESTED** [review]
tokens need to be filtered

File: packages/config/src/layer2s/optimism.ts (hunk 8 above)
Comment (maciekzygmunt) [reply packages/config/src/layer2s/optimism.ts:58]: Via script

File: packages/backend/src/modules/tvl/ChainTvlSubmodule.ts [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+  chainTvlConfig: ChainTvlConfig | false,
+  tokens: Token[],
+  multicallConfig: MulticallConfigEntry[],
+  db: TvlDatabase,
+  priceUpdater: PriceUpdater,
+  http: HttpClient,
+  clock: Clock,
+  logger: Logger,
+): TvlSubmodule | undefined {
+  const name = `${ChainId.getName(chainId)}TvlModule`
```
Comment (antooni) [inline packages/backend/src/modules/tvl/ChainTvlSubmodule.ts:30]: Capitalize the first letter and extract this logic to a function

**Review by antooni — APPROVED** [review]

File: packages/backend/src/modules/tvl/ChainTvlSubmodule.ts (hunk 13 above)
Comment (maciekzygmunt) [reply packages/backend/src/modules/tvl/ChainTvlSubmodule.ts]: [4b4efe5](https://github.com/l2beat/l2beat/pull/1889/commits/4b4efe527e7be0a05cd9eb5fcdcfe49e2ce59dae)

**Pull request merged**